commons-numbers-rootfinder/src/main/java/org/apache/commons/numbers/rootfinder/BrentSolver.java

```diff
@@ -0,0 +1,246 @@
+                           double max) {
+        return findRoot(func, min, 0.5 * (min + max), max);
+    }
+
+    /**
+     * Search the function's zero within the given interval,
+     * starting from the given estimate.
+     *
+     * @param func Function to solve.
+     * @param min Lower bound.
+     * @param initial Initial guess.
+     * @param max Upper bound.
+     * @return the root.
+     * @throws IllegalArgumentException if {@code min > max} or
+     * {@code initial} is not in the {@code [min, max]} interval.
+     * @throws IllegalArgumentException if the given interval does
+     * not bracket the root.
+     */
+    public double findRoot(DoubleUnaryOperator func,
+                           double min,
+                           double initial,
+                           double max) {
+        if (min > max) {
+            throw new SolverException(SolverException.TOO_LARGE, min, max);
+        }
+        if (initial < min ||
+            initial > max) {
+            throw new SolverException(SolverException.OUT_OF_RANGE, initial, min, max);
+        }
+
+        // Return the initial guess if it is good enough.
+        final double yInitial = func.applyAsDouble(initial);
+        if (Math.abs(yInitial) <= functionValueAccuracy) {
+            return initial;
+        }
+
+        // Return the first endpoint if it is good enough.
+        final double yMin = func.applyAsDouble(min);
+        if (Math.abs(yMin) <= functionValueAccuracy) {
+            return min;
+        }
+
+        // Reduce interval if min and initial bracket the root.
+        if (Double.compare(yInitial * yMin, 0.0) < 0) {
+            return brent(func, min, initial, yMin, yInitial);
+        }
+
+        // Return the second endpoint if it is good enough.
+        final double yMax = func.applyAsDouble(max);
+        if (Math.abs(yMax) <= functionValueAccuracy) {
+            return max;
+        }
+
+        // Reduce interval if initial and max bracket the root.
+        if (Double.compare(yInitial * yMax, 0.0) < 0) {
+            return brent(func, initial, max, yInitial, yMax);
+        }
+
+        throw new SolverException(SolverException.BRACKETING, min, yMin, max, yMax);
+    }
+
+    /**
+     * Search for a zero inside the provided interval.
+     * This implementation is based on the algorithm described at page 58 of
+     * the book
+     * <blockquote>
+     *  <b>Algorithms for Minimization Without Derivatives</b>,
+     *  <i>Richard P. Brent</i>,
+     *  Dover 0-486-41998-3
+     * </blockquote>
+     *
+     * @param func Function to solve.
+     * @param lo Lower bound of the search interval.
+     * @param hi Higher bound of the search interval.
+     * @param fLo Function value at the lower bound of the search interval.
+     * @param fHi Function value at the higher bound of the search interval.
+     * @return the value where the function is zero.
+     */
+    private double brent(DoubleUnaryOperator func,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.